### PR TITLE
bugfix - #4789 AWS Secrets Manager - retrieve json secret as a map

### LINKED ISSUE #4789: AWS Secrets Manager - retrieve json secret as a map
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Secrets Manager can store secrets as a plain string or json with flat string to string mapping.

It would be nice to have native support for _retrieving_ secrets stored as json. Currently Terraform returns only string output with no easy way to convert it to map. Ideally, there could be something like `secret_map` attribute on `aws_secretsmanager_secret_version` data source, returning secret as a map.

Terraform probably should attempt to parse the secret string only when `secret_map` attribute is accessed.

The motivation is to conserve the number of secrets (AWS cost), Terraform objects, API calls, etc.

### New or Affected Resource(s)

* data.aws_secretsmanager_secret_version

### Potential Terraform Configuration

```hcl
data "aws_secretsmanager_secret_version" "myapp" {
  secret_id = "${data.aws_secretsmanager_secret.myapp.id}"
}

resource "aws_opsworks_application" "myapp" {

  environment = {
    key    = "SOME_SECRET"
    value  = "${data.aws_secretsmanager_secret_version.myapp.secret_map["SOME_SECRET"]}"
    secure = true
  }

  environment = {
    key    = "OTHER_SECRET"
    value  = "${data.aws_secretsmanager_secret_version.myapp.secret_map["OTHER_SECRET"]}"
    secure = true
  }

}
```

### References

* #4611 - this ticket provides a workaround for _storing_ secrets as json.
* https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_version.html#attributes-reference

### PR DESCRIPTION
**Pull request by ozbillwang**
Fixes #4789

Changes proposed in this pull request:

* AWS Secrets Manager - retrieve json secret as a map (original it is string)

Tasks:
* [x] bugfix
* [ ] test case
* [ ] document updates.

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSAvailabilityZones'

...
```

Here is my test code

```
provider "aws" {
  region = "ap-southeast-2"
}

variable "secret" {
  type = "map"
  default = {
    name = "team_project_dev"
    key = "db_master_password"
  }
}

data "aws_secretsmanager_secret" "secret" {
  name = "${lookup(var.secret, "name")}"
}

data "aws_secretsmanager_secret_version" "secret" {
  secret_id = "${data.aws_secretsmanager_secret.secret.id}"
}

locals {
  master_password = "${lookup(data.aws_secretsmanager_secret_version.secret.secret_string, lookup(var.secret, "key"))}"
}

output "master_password" {
  value = "${local.master_password}"
}
```
Result:

```
$ terraform apply 

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

master_password = test
```

### CODE REVIEW

File: aws/data_source_aws_secretsmanager_secret_version.go [hunk 1]
```diff
@@ -20,7 +21,7 @@ func dataSourceAwsSecretsManagerSecretVersion() *schema.Resource {
 				ForceNew: true,
 			},
 			"secret_string": {
-				Type:      schema.TypeString,
```
Comment (bflad) [inline aws/data_source_aws_secretsmanager_secret_version.go]: Changing an attribute type is a breaking change for existing Terraform configurations. This new behavior should be handled by creating a new attribute and a new acceptance test added to cover if and when this new attribute is available.

There are also other potential issues with introducing this second attribute as briefly outlined in: https://github.com/terraform-providers/terraform-provider-aws/pull/4613#issuecomment-391337108

File: aws/data_source_aws_secretsmanager_secret_version.go [hunk 2]
```diff
@@ -76,7 +77,11 @@ func dataSourceAwsSecretsManagerSecretVersionRead(d *schema.ResourceData, meta i
 
 	d.SetId(fmt.Sprintf("%s|%s", secretID, version))
 	d.Set("secret_id", secretID)
-	d.Set("secret_string", output.SecretString)
+	var sec map[string]interface{}
+	if err = json.Unmarshal([]byte(*output.SecretString), &sec); err != nil {
+		return fmt.Errorf("Error getting secret string: %s", err)
```
Comment (bflad) [inline aws/data_source_aws_secretsmanager_secret_version.go]: It is valid to create a Secrets Manager secret that is a single string value (not JSON). This error should only be logged to not break handling with those types of secrets.

**Comment by bflad** [comment]
I provided additional information about this handling in https://github.com/terraform-providers/terraform-provider-aws/issues/4789#issuecomment-402767316

**Comment by ozbillwang** [comment]
@bflad 

Thanks for reviewing, this PR is not suitable, I will close it and wait for new v0.12 release.

**Pull request closed without merging**